rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -0,0 +1,484 @@
+    /// * `timeout` - The duration until the Client stops waiting for a response to the request, it is rounded up to the nearest second.
+    ///
+    /// Returns the updated [`NotificationResponse`].
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument)
+    /// if the `timeout` is zero or > `u32::max`, or there is an error building the request.
+    ///
+    /// [`struct@Error`] of kind [`SerializationError`](ErrorKind::SerializationError)
+    /// if there is an error serializing the request.
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError)
+    /// if there is an error returned by the ADR Service.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError)
+    /// if there are any underlying errors from the AIO RPC protocol.
+    pub async fn notify_asset_update(
+        &self,
+        asset_name: String,
+        notification_type: bool,
+        timeout: Duration,
+    ) -> Result<NotificationResponse, AzureDeviceRegistryError> {
+        let notification_paylaod = NotifyOnAssetUpdateRequestPayload {
+            notification_request: NotifyOnAssetUpdateRequestSchema {
+                asset_name,
+                notification_message_type: if notification_type {
+                    NotificationMessageType::On
+                } else {
+                    NotificationMessageType::Off
+                },
+            },
+        };
+
+        let command_request = rpc_command::invoker::RequestBuilder::default()
+            .payload(notification_paylaod)
+            .map_err(|e| AzureDeviceRegistryError(ErrorKind::SerializationError(e.to_string())))?
+            .timeout(timeout)
+            .build()
+            .map_err(|e| AzureDeviceRegistryError(ErrorKind::InvalidArgument(e.to_string())))?;
+
+        let result = self
+            .notify_on_asset_update_command_invoker
+            .invoke(command_request)
+            .await;
+
+        match result {
+            Ok(response) => Ok(response.payload.notification_response),
+            Err(e) => Err(AzureDeviceRegistryError(ErrorKind::from(
+                ErrorKind::AIOProtocolError(e),
+            ))),
+        }
+    }
+
+    /// Shutdown the [`Client`]. Shuts down the underlying command invokers for get and put operations.
+    ///
+    /// Note: If this method is called, the [`Client`] should not be used again.
+    /// If the method returns an error, it may be called again to re-attempt unsubscribing.
+    ///
+    /// Returns Ok(()) on success, otherwise returns [`struct@Error`].
+    /// # Errors
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError)
+    /// if the unsubscribe fails or if the unsuback reason code doesn't indicate success.
+    pub async fn shutdown(&self) -> Result<(), AzureDeviceRegistryError> {
+        self.get_asset_endpoint_profile_command_invoker
```

Comment:
with this many invokers, we might want to have these all attempt to run and return any errors instead of having it break out of the rest of them if one fails.